### ISSUE TITLE
Formatter options

### ISSUE THREAD

**Issue by mathieumg** [eslint/eslint#2989]
Would you be open to a PR that allows to provide options for custom formatters as a second CLI argument and if so, how should it be named and what form should it take? (`-fopt`,`--format-option`?)

I think there is a need for that when you see things like https://github.com/ganimomer/eslint-stats#available-reporters where every time you want to add a formatting option, you need to double the amount of "formatter files" you have to allow for multiple option combination. (And that only works for boolean options, even!)

I've come across the same type of issue in https://github.com/royriojas/eslint-friendly-formatter/issues/7

Thank you!

**Comment by royriojas** [comment]
+1 I end using environment variables for this kind of stuff but it feels more like a hack rather than a proper solution. Very fragile.

**Comment by nzakas** [comment]
Can you explain a use case? Part of the limitations on formatters is to prevent them from doing too much.

**Comment by royriojas** [comment]
for example:

- [passing a flag to disable colored output](https://github.com/royriojas/eslint-friendly-formatter/pull/2)
- [passing a flag to report the path to the files as absolute paths vs relative ones](https://github.com/royriojas/eslint-friendly-formatter/pull/7)

Those two were things I had to add environment flags to support them.

But maybe you're right, limiting what reporters can do might be a good thing.

**Comment by nzakas** [comment]
Yeah, I'm having a hard time rationalizing adding this capability right now. You can always extract common functionality into a single file and extend that to create multiple formatters. I think that's the way to go.

**Comment by mathieumg** [comment]
> You can always extract common functionality into a single file and extend that to create multiple formatters.

That's not the troublesome part, the problem is to inform the formatter of our intent. The more options are added, the more "sub-formatters" will be needed (even if there aren't that many options), every new boolean option doubles the amount of named js files we would need to use the different combinations. (See my first linked example in my original post) Since we pass a filepath to ESLint, we do not really have any way (that I see, I'd be happy to be wrong here) to send options, hence the desire to introduce one.

Another way to "transmit" information to the formatter would be using a query string that ESLint would extract, but that feels both unnatural for a filesystem, and too complex for its own good. I think simply introducing a `--format-option` (that can be used like `--format-option myboolean` or `--format-option myvar=123`, and repeated if there are many options) would be more than enough to cover the ground cases.

**Comment by ilyavolodin** [comment]
Hmm.. I wonder if formatters should be added as exportable object to `eslint-plugin-` infrastructure? It feels like that would solve all of the problems described above, plus add ability to share formatters easily by publishing them to NPM.

**Comment by mathieumg** [comment]
I'd be fine with that too, anything that improves the situation I described above. :+1:

**Comment by ilyavolodin** [comment]
Opened #3013 to discuss formatters in plugins.

**Comment by nzakas** [comment]
@mathieumg again, formatters weren't designed to be that complicated. There is an intentional constraint here to keep them as simple as possible. Formatters are just not intended to have multiple options and I don't see a strong argument why that should change. It seems like you are still able to do what you want to do and the end result is a better user experience by forcing separate named formatters.

**Comment by royriojas** [comment]
@nzakas could you consider adding an option to just spit to the std output the unformatted results? I guess that will solve all of the issues related to the formatters that way some one can just use that as the input for a custom formatter that does accept extra parameters.
some like this

```bash
eslint --raw ./src/**/*.*.js | formatter --no-color --full-paths
```

Sorry for keep coming back on the issue. I do see the value on keeping things simple in the formatters, so if you say there is no way to do something like that, then it is fine.

Thanks for so useful tool
